src/modules/alwaysontop/AlwaysOnTop/FrameDrawer.cpp

```diff
@@ -0,0 +1,205 @@
+#include "pch.h"
+#include "FrameDrawer.h"
+
+std::unique_ptr<FrameDrawer> FrameDrawer::Create(HWND window)
+{
+    auto self = std::make_unique<FrameDrawer>(window);
+    if (self->Init())
+    {
+        return self;
+    }
+
+    return nullptr;
+}
+
+FrameDrawer::FrameDrawer(FrameDrawer&& other) :
+    m_window(other.m_window), 
+    m_clientRect(std::move(other.m_clientRect)), 
+    m_renderTarget(std::move(other.m_renderTarget)), 
+    m_sceneRect(std::move(other.m_sceneRect)), 
+    m_shouldRender(other.m_shouldRender.load()), 
+    m_abortThread(other.m_abortThread.load()), 
+    m_renderThread(std::move(m_renderThread))
+{
+}
+
+FrameDrawer::FrameDrawer(HWND window) :
+    m_window(window), m_renderTarget(nullptr), m_shouldRender(false)
+{
+}
+
+FrameDrawer::~FrameDrawer()
+{
+    {
+        std::unique_lock lock(m_mutex);
+        m_abortThread = true;
+        m_shouldRender = true;
+    }
+    m_cv.notify_all();
+    m_renderThread.join();
+
+    if (m_renderTarget)
+    {
+        m_renderTarget->Release();
+    }
+}
+
+bool FrameDrawer::Init()
+{
+    // Obtain the size of the drawing area.
+    if (!GetClientRect(m_window, &m_clientRect))
+    {
+        return false;
+    }
+
+    HRESULT hr;
+
+    // Create a Direct2D render target
+    // We should always use the DPI value of 96 since we're running in DPI aware mode
+    auto renderTargetProperties = D2D1::RenderTargetProperties(
+        D2D1_RENDER_TARGET_TYPE_DEFAULT,
+        D2D1::PixelFormat(DXGI_FORMAT_UNKNOWN, D2D1_ALPHA_MODE_PREMULTIPLIED),
+        96.f,
+        96.f);
+
+    auto renderTargetSize = D2D1::SizeU(m_clientRect.right - m_clientRect.left, m_clientRect.bottom - m_clientRect.top);
+    auto hwndRenderTargetProperties = D2D1::HwndRenderTargetProperties(m_window, renderTargetSize);
+
+    hr = GetD2DFactory()->CreateHwndRenderTarget(renderTargetProperties, hwndRenderTargetProperties, &m_renderTarget);
+
+    if (!SUCCEEDED(hr))
+    {
+        return false;
+    }
+
+    m_renderThread = std::thread([this]() { RenderLoop(); });
+
+    return true;
+}
+
+void FrameDrawer::Hide()
+{
+    bool shouldHideWindow = true;
+    {
+        std::unique_lock lock(m_mutex);
+        shouldHideWindow = m_shouldRender;
```

Comment:
is this `shouldHideWindow` part necessary ? Why not hide unconditionally ?

Reply:
You're right, I'll clean this up.